cirq/qis/measures.py

```diff
@@ -69,3 +74,53 @@ def von_neumann_entropy(density_matrix: np.ndarray) -> float:
     """
     eigenvalues = np.linalg.eigvalsh(density_matrix)
     return scipy.stats.entropy(abs(eigenvalues), base=2)
+
+
+def process_fidelity(clean_circuit, noisy_circuit, qubits) -> float:
+    """Calculates the average fidelity of a noisy circuit.
+
+    The code uses the Kraus representation for open circuits, when decomposing
+    into noisy channels. The formula for process fidelity can be found at
+    equation (2) of "Quantum Gate Fidelity in Terms of Choi Matrices" by
+    Nathaniel Johnston and David W. Kribs which can be found at:
+    https://arxiv.org/pdf/1102.0948.pdf
+
+    Another useful reference is "A simple formula for the average gate fidelity
+    of a quantum dynamical operation" by Michael A. Nielsen which can be found
+    at:
+    https://arxiv.org/pdf/quant-ph/0205035.pdf
+
+    Args:
+        clean_circuit: The perfect circuit (no noise, closed).
+        noisy_circuit: The circuit with noise gates (open circuit).
+        qubits: The list of qubits.
+    Returns:
+        A scalar that is the average (process) entropy
+    """
+    n = len(qubits)
+    d = 2**n
+    qubit_map = {q.with_dimension(1): i for i, q in enumerate(qubits)}
+
+    kraus_operations = [clean_circuit.unitary().reshape([2] * (2 * n))]
+    for op in noisy_circuit.all_operations():
+        target_axes = [qubit_map[q.with_dimension(1)] for q in op.qubits]
+
+        next_kraus_operations = []
+
+        for op_kraus in protocols.channel(op, default=None):
```

Comment:
Am I right in thinking of this loop as basically a new noisy simulator? This should likely be factored out. Of course @viathor will not hesitate to point out that we need to nail down the story on our representations for channels. I know our existing noise channels describe themselves with Kraus operators, but as you can see here: they get a little clunky to compose

Reply:
+1 to both factoring this out (good to do in this PR) and representation (out of scope for this PR).

Also, I suppose "a little clunky" is a euphemism for "exponentially clunky" ;-)

Reply:
Could we change this to a pure linalg function like `compose_kraus_operations`, that takes the initial density matrix and like `operations: Iterable[Tuple[Sequence[ndarray], Sequence[int]]]`, (i.e. a list of (kraus, axes) tuples)? Then this would be a useful general function, not bound to our definition of a circuit. We could have another function that calculates the fidelity of that result.

If we have those, then we don't necessarily need to "standardize" on a representation. We can leave this as pure linalg, and anyone can use it by writing a couple lines to extract the (kraus, axes) list from their noisy circuit and feed it into here. (Maybe we offer that extraction as a separate utility function elsewhere).